torch/csrc/DataLoader.cpp

```diff
@@ -0,0 +1,253 @@
+  // from this thread
+  sigset_t sigset, old_sigset;
+  sigemptyset(&sigset);
+  sigaddset(&sigset, SIGCHLD);
+  if (sigprocmask(SIG_BLOCK, &sigset, &old_sigset) != 0) {
+    return -1;
+  }
+  // Acquire ``lock'' so handlers on other threads can't interrupt
+  char c;
+  read(comm_pipe[0], &c, 1);
+
+  pid_t *old_ptr = worker_pids;
+  if (new_size < num_worker_pids) {
+    num_worker_pids = new_size;
+    worker_pids = new_ptr;
+  } else {
+    worker_pids = new_ptr;
+    num_worker_pids = new_size;
+  }
+  free(old_ptr);
+
+  // Release ``lock''
+  write(comm_pipe[1], &c, 1);
+  // Restore handler for this thread.
+  if (sigprocmask(SIG_SETMASK, &old_sigset, NULL) != 0) {
+    return -1;
+  }
+  return 0;
+}
+
+static struct sigaction orig_SIGCHLD_sa;
+
+// SIGCHLD hander should be registered on main loader process to catch any
+// worker failing.
+// Python handles are _set_main_signal_handers_for_workers() and
+// _remove_main_signal_handers_for_workers().
+static void handler_SIGCHLD_main(int sig, siginfo_t *info, void *ctx) {
+  // Acquire ``lock'' so make sure that worker_pids won't change
+  char c;
+  read(comm_pipe[0], &c, 1);
+
+  int error;
+  siginfo_t infop;
+
+  // Only check the pids we care about so that Python can see other processes'
+  // status.
+  for (size_t i = 0; i < num_worker_pids; i++) {
+    // Use waitid rather than waitpid so that we can set NOWAIT, and that Python
+    // can get whatever info it wants about the child process.
+    error = waitid(P_PID, worker_pids[i], &infop, WEXITED|WNOHANG|WNOWAIT);
+    if (error < 0)  // ignore errors
+      continue;
+    if ((infop.si_code == CLD_EXITED && infop.si_status != 0) ||  // exit with error
+        (infop.si_code == CLD_KILLED) ||
+        (infop.si_code == CLD_DUMPED) ||
+        (infop.si_code == CLD_TRAPPED)) {
+      _exit(EXIT_FAILURE);
+    }
+  }
+
+  // Release ``lock''
+  write(comm_pipe[1], &c, 1);
+
+  // Call the overridden handler.
+  if ((orig_SIGCHLD_sa.sa_flags | SA_SIGINFO) != 0) {
+    // handler is sa_sigaction, this shouldn't really be SIG_IGN or SIG_DFL, but
+    // they happen to be a union, and this fact is apparently used in Python,
+    // so check here.
+    if (orig_SIGCHLD_sa.sa_sigaction == (void (*)(int, siginfo_t *, void *)) SIG_IGN) {
+      // SIG_IGN for SIGCHLD is to reap the child and do nothing else.
+      while (waitpid(-1, 0, WNOHANG) > 0) {}
+    } else if (orig_SIGCHLD_sa.sa_sigaction != (void (*)(int, siginfo_t *, void *)) SIG_DFL) {
+      // SIG_DFL for SIGCHLD is to leave the child as a zombie (do nothing)
+      orig_SIGCHLD_sa.sa_sigaction(sig, info, ctx);
+    }
+  } else {
+    // handler is sa_handler
+    // https://stackoverflow.com/a/24080440
+    if (orig_SIGCHLD_sa.sa_handler == SIG_IGN) {
+      while (waitpid(-1, 0, WNOHANG) > 0) {}
+    } else if (orig_SIGCHLD_sa.sa_handler != SIG_DFL) {
+      orig_SIGCHLD_sa.sa_handler(sig);
+    }
+  }
+}
+
+// returns -1 on error.
+static int isSIGCHLDHanderSet() {
```

Comment:
I really thing we shouldn't follow the C error checking conventions in our C++ code. It's unnecessarily complicated and exceptions make these things a lot easier. Just raise an error if an error happens, and return a boolean value otherwise